loader/src/main/java/net/neoforged/fml/common/asm/enumextension/RuntimeEnumExtender.java

```diff
@@ -0,0 +1,435 @@
+import net.neoforged.neoforgespi.language.IModInfo;
+import org.jetbrains.annotations.ApiStatus;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.GeneratorAdapter;
+import org.objectweb.asm.commons.Method;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.AnnotationNode;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.FieldInsnNode;
+import org.objectweb.asm.tree.FieldNode;
+import org.objectweb.asm.tree.InsnList;
+import org.objectweb.asm.tree.InsnNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+
+/**
+ * Transforms enums implementing {@link IExtensibleEnum} to add additional entries loaded from files provided by mods
+ */
+@ApiStatus.Internal
+public class RuntimeEnumExtender implements ILaunchPluginService {
+    private static final EnumSet<Phase> YAY = EnumSet.of(Phase.BEFORE);
+    private static final EnumSet<Phase> NAY = EnumSet.noneOf(Phase.class);
+    private static final Type MARKER_IFACE = Type.getType(IExtensibleEnum.class);
+    private static final Type INDEXED_ANNOTATION = Type.getType(IndexedEnum.class);
+    private static final Type NAMED_ANNOTATION = Type.getType(NamedEnum.class);
+    private static final Type RESERVED_ANNOTATION = Type.getType(ReservedConstructor.class);
+    private static final Type ENUM_PROXY = Type.getType(EnumProxy.class);
+    private static final Type NET_CHECK = Type.getType(NetworkedEnum.NetworkCheck.class);
+    private static final Type EXT_INFO = Type.getType(ExtensionInfo.class);
+    private static final String EXT_INFO_GETTER_DESC = Type.getMethodDescriptor(EXT_INFO);
+    private static final String EXT_INFO_CTOR_DESC = Type.getMethodDescriptor(
+            Type.VOID_TYPE, Type.BOOLEAN_TYPE, Type.INT_TYPE, Type.INT_TYPE, NET_CHECK);
+    private static final Type NETWORKED_ANNOTATION = Type.getType(NetworkedEnum.class);
+    private static final Type EXTENDER = Type.getType(RuntimeEnumExtender.class);
+    private static final Type ARRAYS = Type.getType("Ljava/util/Arrays;");
+    private static final int ENUM_FLAGS = Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC | Opcodes.ACC_FINAL | Opcodes.ACC_ENUM;
+    private static final int EXT_INFO_FLAGS = Opcodes.ACC_PRIVATE | Opcodes.ACC_STATIC | Opcodes.ACC_FINAL;
+    private static volatile Map<String, List<EnumPrototype>> prototypes = Map.of();
+
+    @Override
+    public String name() {
+        return "runtime_enum_extender";
+    }
+
+    @Override
+    public EnumSet<Phase> handlesClass(Type classType, boolean isEmpty) {
+        return isEmpty || !prototypes.containsKey(classType.getInternalName()) ? NAY : YAY;
+    }
+
+    @Override
+    public boolean processClass(final Phase phase, final ClassNode classNode, final Type classType) {
+        if ((classNode.access & Opcodes.ACC_ENUM) == 0 || !classNode.interfaces.contains(MARKER_IFACE.getInternalName())) {
```

Comment:
Shouldn't it be an error to try to extend a non-enum class?

Reply:
That's a fair point, yeah. I'll add a throw if the target is not an enum.